### PR TITLE
ported !urban to pybot

### PR DESCRIPTION
**Pull request by nicklambourne**
Ported !urban functionality to pybot.

screen shot 2018-06-16 at 1 17 27 pm

screen shot 2018-06-16 at 1 17 36 pm

screen shot 2018-06-16 at 1 18 50 pm

### CODE REVIEW

File: uqcsbot/scripts/urban.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    # If search phrase is not found, notify user.
+    if results['result_type'] != 'exact':
+        bot.post_message(command.channel, '> No results found for ' + search_term + '. ¯\\_(ツ)_/¯')
+        return
+
+    # Get best definition (by most 'thumbs up') and construct response.
+    sorted_definitions = sorted(results['list'], key=lambda definition: definition['thumbs_up'], reverse=True)
+    best_definition = sorted_definitions[0]
+    response_example = best_definition.get('example', '')
+    more_definitions = '_ more definitions at ' + http_response.url.replace(URBAN_API_ENDPOINT, URBAN_USER_ENDPOINT) +\
```
Comment (mitchmcdee) [inline uqcsbot/scripts/urban.py]: add `+` on next line :^) easier to read

Comment (mitchmcdee) [reply uqcsbot/scripts/urban.py]: nice work with api endpoint / user endpoint swapping 👍

File: uqcsbot/scripts/urban.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+                       ' _' if len(sorted_definitions) > 1 else ''
```
Comment (mitchmcdee) [inline uqcsbot/scripts/urban.py]: what's this accomplishing exactly? looks a lil messy

File: uqcsbot/scripts/urban.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+    # Format and send response to user in channel query was sent from.
+    bot.post_message(command.channel, '*' + search_term.title() + '* \n ' +
```
Comment (mitchmcdee) [inline uqcsbot/scripts/urban.py]: given the non-trivial formatting, I'd abstract it out to its own variable to make it a bit easier to read

File: uqcsbot/scripts/urban.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+                     best_definition['definition'].capitalize() +
+                     '\n> ' + response_example.replace('\r\n', '\n> ') + '\n' + more_definitions)
```
Comment (mitchmcdee) [inline uqcsbot/scripts/urban.py]: this formatting seems messy with the replacing and arrows n shit, can you explain why they're necessary and/or try to simplify it?

File: uqcsbot/scripts/urban.py (hunk 3 above)
Comment (TRManderson) [reply uqcsbot/scripts/urban.py]: A lot of the `+` concatenation could also be replaced with an f-string

**Comment by TRManderson** [comment]
Can you delete the JS version from `unimplemented_commands` in this PR too?

**Comment by mitchmcdee** [comment]
Not sure if still relevant, but your screenshot has a bug bahahahaha 😂double `??` on url is invalid, was that fixed?

File: uqcsbot/scripts/urban.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+    # If search phrase is not found, notify user.
+    if results['result_type'] != 'exact':
+        bot.post_message(command.channel, f'> No results found for {search_term}. ¯\\_(ツ)_/¯')
+        return
+
+    # Get best definition (by most 'thumbs up') and construct response.
+    sorted_definitions = sorted(results['list'], key=lambda definition: definition['thumbs_up'], reverse=True)
+    best_definition = sorted_definitions[0]
+    response_example = best_definition.get('example', '').replace('\r\n', '\n> ')  # Format example with block quotes.
```
Comment (mitchmcdee) [inline uqcsbot/scripts/urban.py]: I finally understand `.replace('\r\n', '\n> ')`, it puts each example in its own block quote. Maybe just have something like:
```
examples = best_definition.get('example', '').split('\r\n')
response_examples = '\n'.join(f'> {example}' for example in examples)
```
might not be as sleek, but infinitely more readable

File: uqcsbot/scripts/urban.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    more_definitions = f'_ more definitions at {http_response.url.replace(URBAN_API_ENDPOINT, URBAN_USER_ENDPOINT)} _'
+
+    # Format message and send response to user in channel query was sent from.
+    message = f'*{search_term.title()}*\n{best_definition["definition"].capitalize()}\n> ' \
```
Comment (mitchmcdee) [inline uqcsbot/scripts/urban.py]: still hard to read and understand, would prefer something like:
```
message = f'*{search_term.title()}*\n' \
         + f'{best_definition['definition'].capitalize()}\n' \
         + '\n'.join(f'> {example}' for example in examples)
if len(sorted_definitions) > 1:
    endpoint_url = http_response.url.replace(URBAN_API_ENDPOINT, URBAN_USER_ENDPOINT)
    message += f'\n_more definitions at {endpoint_url}_'
```

Comment (mitchmcdee) [reply uqcsbot/scripts/urban.py]: note I used the joint examples thing from above, so you could probably remove that and have it like:
```
examples = best_definition.get('example', '').split('\r\n')
message = f'*{search_term.title()}*\n' \
         + f'{best_definition['definition'].capitalize()}\n' \
         + '\n'.join(f'> {example}' for example in examples)
if len(sorted_definitions) > 1:
    endpoint_url = http_response.url.replace(URBAN_API_ENDPOINT, URBAN_USER_ENDPOINT)
    message += f'\n_more definitions at {endpoint_url}_'
```

Comment (mitchmcdee) [reply uqcsbot/scripts/urban.py]: you get the gist
anyway up to you-- just don't be afraid to decondense ya code for the sake of readability ;)

File: uqcsbot/scripts/urban.py [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+        return
+    results = http_response.json()
+
+    # If search phrase is not found, notify user.
+    if results['result_type'] != 'exact':
+        bot.post_message(command.channel, f'> No results found for {search_term}. ¯\\_(ツ)_/¯')
+        return
+
+    # Get best definition (by most 'thumbs up') and construct response.
+    sorted_definitions = sorted(results['list'], key=lambda definition: definition['thumbs_up'], reverse=True)
```
Comment (mitchmcdee) [inline uqcsbot/scripts/urban.py]: can probably shorten to just `def` to trim down this line length

File: uqcsbot/scripts/urban.py (hunk 5 above)
Comment (nicklambourne) [reply uqcsbot/scripts/urban.py]: It's just the one example, but I had a choice between formatting it with '>>>' and sending the more examples part as a second message (in order for it to not be in the block quote), or prepending '>' on each line of the example.

Comment (nicklambourne) [reply uqcsbot/scripts/urban.py]: That would still work though. 😁

**Pull request merged**